### PR TITLE
API documentation: h2d package

### PR DESCRIPTION
**Pull request by Yanrishatum**
This PR is done thanks to donations from 01010111, Kenny Goff, Weast, Ryan Rohrer and Daniel Sun.

I took on an effort of fixing one of long-standing issues community had with Heaps: Little to no documentation. Because I won't be able to cover everything in one sweep, it will be done in parts, and this is h2d package part. 

While current docs are a draft version and not the final one (tho I would say majority of it would be left untouched), I would welcome anyone to provide commentary on unclear moments, points out phrasing issues, grammar errors and such.

I did not touch documentation of the following classes:
* ~`h2d.Camera` - because new camera PR. :) Why would I touch that.~ Merged already and is documented.
* `h2d.CdbLevel` - it's an external dependency not many use, documented to a degree and majority of it's API is rather self-evident.
* `h2d.Layers` - because it's already documented.
* ~`h2d.KeyFrames` - as with CdbLevel it's an external dependency.~ Partially documented.
* ~`h2d.filter.Ambient` - because I couldn't figure out what exactly it supposed to do.~ Thanks to Cannasse for explanation.

# Current state of PR

* [x] Cover root `h2d` package classes
* [x] Cover `h2d.col` package
*  Cover `h2d.domkit` package~
* [x] Cover `h2d.filter` package
* [x] Go trough entire PR and add documentation to private methods that are meant to be overridden with usage of `@:dox(show)`
* [x] Go trough entire PR and redraft it. Fix typos, unclear statements and in general make it more readable.

Temporary hosting of the WIP docs can be found here: https://wb.yanrishatum.ru/heaps/api/

### CODE REVIEW

File: h2d/Console.hx [hunk 1]
```diff
@@ -179,17 +228,26 @@ class Console #if !macro extends h2d.Object #end {
...
+	**/
 	public function hide() {
 		bg.visible = false;
 		tf.text = "";
 		hintTxt.text = "";
 		tf.cursorIndex = -1;
 	}
 
+	/**
+		Hides the shows and focuses the Console.
```
Comment (dpomier) [inline h2d/Console.hx:249]:
```suggestion
		Shows and focuses the Console.
```

File: h2d/Drawable.hx [hunk 2]
```diff
@@ -13,12 +13,14 @@ class Drawable extends Object {
 
 	/**
 		By enabling smoothing, scaling the object up or down will use hardware bilinear filtering resulting in less crisp aspect.
```
Comment (dpomier) [inline h2d/Drawable.hx:15]:
```suggestion
		By enabling smoothing, scaling the object up or down will use hardware bilinear filtering resulting in a less crisp aspect.
```

File: h2d/Drawable.hx [hunk 3]
```diff
@@ -13,12 +13,14 @@ class Drawable extends Object {
...
-		By default smooth is null and then Scene.defaultSmooth value is used.
+		By default smooth is `null` in which case `Scene.defaultSmooth` value is used.
 	**/
 	public var smooth : Null<Bool>;
 
 	/**
-		By enabling tile wrapping, you can have tiles which size exceed the texture size and will repeat instead of displaying clamped pixels.
+		By enabling tile wrapping, you can have tiles which size exceed the texture size and will repeat instead of displaying clamped pixels.  
+		tileWrap does not wrap within `Tile` bounds but withing underlying `Texture` and due to that, rendering a region of a texture with
+		tileWrap enabled won't cause that region to be tiled, but will extend into out-of-tile-bounds areas.
```
Comment (dpomier) [inline h2d/Drawable.hx:23]:
```suggestion
		`tileWrap` does not wrap within `Tile` bounds but withing underlying `Texture` and due to that, rendering a region of a texture with
		`tileWrap` enabled won't cause that region to be tiled, but will extend into out-of-tile-bounds areas.
```

File: h2d/Drawable.hx [hunk 4]
```diff
@@ -91,7 +93,8 @@ class Drawable extends Object {
 	}
 
 	/**
-		Set the `colorMatrix` value by specifying which effects to apply. Calling adjustColor() reset the colorMatrix to `null`
+		Set the `colorMatrix` value by specifying which effects to apply.
+		Calling `adjustColor()` without arguments will reset the colorMatrix to `null`
```
Comment (dpomier) [inline h2d/Drawable.hx:97]:
```suggestion
		Calling `adjustColor()` without arguments will reset the colorMatrix to `null`.
```

File: h2d/Drawable.hx [hunk 5]
```diff
@@ -164,7 +167,8 @@ class Drawable extends Object {
 	}
 
 	/**
-		Add a shader to the object shaders
+		Add a shader to the object shaders
```
Comment (dpomier) [inline h2d/Drawable.hx:170]:
```suggestion
		Add a shader to the object shaders.
```

File: h2d/Dropdown.hx [hunk 6]
```diff
@@ -30,22 +30,54 @@ private class Fake extends Object {
 	}
 }
 
+/**
+	`h2d.Dropdown` is a simple UI component that creates an interactive drop-down list.
+	Dropdown will add an `h2d.Flow` to `Scene` when opening in order to be visible above other objects. See `dropdownLayer`.  
+	Note that when dropdownList opens and closes, item objects will recieve the `onHierarchyChanged` callback.
```
Comment (dpomier) [inline h2d/Dropdown.hx:36]:
```suggestion
	Note that when `dropdownList` opens and closes, item objects will receive the `onHierarchyChanged` callback.
```

File: h2d/Dropdown.hx [hunk 7]
```diff
@@ -30,22 +30,54 @@ private class Fake extends Object {
...
+**/
 class Dropdown extends Flow {
 	var items : Array<h2d.Object>;
 	var fake : Fake;
 	var cursor : h2d.Bitmap;
 	var arrow : h2d.Bitmap;
 
+	/**
+		A background Tile that is shown when user hover over an item in the dropdown list.
```
Comment (dpomier) [inline h2d/Dropdown.hx:45]:
```suggestion
		A background Tile that is shown when the user overs an item in the dropdown list.
```

File: h2d/Dropdown.hx [hunk 8]
```diff
@@ -30,22 +30,54 @@ private class Fake extends Object {
...
+	**/
 	public var tileOverItem(default, set) : h2d.Tile;
+	/**
+		A Tile used to visualize an arrow of the dropdown when list is closed.
```
Comment (dpomier) [inline h2d/Dropdown.hx:49]:
```suggestion
		A Tile used to visualize an arrow of the dropdown when the list is closed.
```

File: h2d/Dropdown.hx [hunk 9]
```diff
@@ -30,22 +30,54 @@ private class Fake extends Object {
...
+	**/
 	public var tileArrow(default, set) : h2d.Tile;
+	/**
+		A Tile used to visualize and arrow of the dropdown when list is open.
```
Comment (dpomier) [inline h2d/Dropdown.hx:53]:
```suggestion
		A Tile used to visualize and arrow of the dropdown when the list is open.
```

File: h2d/Dropdown.hx [hunk 10]
```diff
@@ -30,22 +30,54 @@ private class Fake extends Object {
...
+	**/
 	public var tileArrowOpen : h2d.Tile;
-
+	
+	/**
+		When disabled, user would not be able to change selected item. ( default : true )
```
Comment (dpomier) [inline h2d/Dropdown.hx:58]:
```suggestion
		When disabled, the user would not be able to change the selected item. ( default : true )
```

File: h2d/Dropdown.hx [hunk 11]
```diff
@@ -30,22 +30,54 @@ private class Fake extends Object {
...
+	**/
 	public var canEdit(default,set) : Bool = true;
+	/**
+		A reference to `h2d.Flow` that will contain items.
+		Adding objects to this Flow will not automatically add them to item list, see `addItem`.
```
Comment (dpomier) [inline h2d/Dropdown.hx:63]:
```suggestion
		Adding objects to this Flow will not automatically add them to the item list, see `addItem`.
```

File: h2d/Dropdown.hx [hunk 12]
```diff
@@ -30,22 +30,54 @@ private class Fake extends Object {
...
+	**/
 	public var dropdownList : Flow;
+	/**
+		A Scene layer to which dropdownList will be added when opening dropdown. ( default : 0 )
```
Comment (dpomier) [inline h2d/Dropdown.hx:67]:
```suggestion
		A Scene layer to which `dropdownList` will be added when opening dropdown. ( default : 0 )
```

File: h2d/Dropdown.hx [hunk 13]
```diff
@@ -30,22 +30,54 @@ private class Fake extends Object {
...
+		Currently selected item index. To deselect an item, set it to `-1`.
+	**/
 	public var selectedItem(default, set) : Int = -1;
+	/**
+		Currently highlighted item index.
+	**/
 	public var highlightedItem(default, null) : Int = -1;
-
+	/**
+		When enabled, dropdown list will appear above the dropdown.
```
Comment (dpomier) [inline h2d/Dropdown.hx:79]:
```suggestion
		When enabled, the dropdown list will appear above the dropdown.
```

File: h2d/Dropdown.hx [hunk 14]
```diff
@@ -157,6 +189,9 @@ class Dropdown extends Flow {
 		return tileOverItem = t;
 	}
 
+	/**
+		Adds an Object to the dropdown list. Object is not restricted to be same type across all items.
```
Comment (dpomier) [inline h2d/Dropdown.hx:193]:
```suggestion
		Adds the Object `s` to the dropdown list. `s` is not restricted to be the same type across all items.
```

File: h2d/Font.hx [hunk 15]
```diff
@@ -77,14 +107,33 @@ enum FontType {
...
 
 class Font {
-
 	public var name(default, null) : String;
+	/**
+		Current font size. Font can be resized with `resizeTo`.
+	**/
 	public var size(default, null) : Int;
+	/**
+		The baseline value of the font represent the base on which characters will sit.
```
Comment (dpomier) [inline h2d/Font.hx:116]:
```suggestion
		The baseline value of the font represents the base on which characters will sit.
```

File: h2d/Font.hx [hunk 16]
```diff
@@ -77,14 +107,33 @@ enum FontType {
...
+		Font line height provides vertical offset for each new line of the text.
+	**/
 	public var lineHeight(default, null) : Float;
+	/**
+		Reference to the source Tile containing all glyphs of the Font.
+	**/
 	public var tile(default,null) : h2d.Tile;
 	public var tilePath(default,null) : String;
+	/**
+		The font type. BitmapFonts rendered as-is, but SDF fonts will use extra shader to produce scalable smooth fonts.
```
Comment (dpomier) [inline h2d/Font.hx:130]:
```suggestion
		The font type. BitmapFonts rendered as-is, but SDF fonts will use an extra shader to produce scalable smooth fonts.
```

File: h2d/Font.hx [hunk 17]
```diff
@@ -77,14 +107,33 @@ enum FontType {
...
+	**/
 	public var type : FontType;
+	/**
+		Font charset allows to resolve specific char codes that are not directly present in glyph map as well as detect spaces.
+		Defaults to `hxd.Charset.getDefault()`
```
Comment (dpomier) [inline h2d/Font.hx:135]:
```suggestion
		Defaults to `hxd.Charset.getDefault()`.
```

**Comment by Yanrishatum** [comment]
@dpomier Accepted some of suggestions (and I think some would be rephrased, as they are a bit unclear), but I'm not sure about overabundance of `the` additions. English is not my primary language, but with all those extra `the` it reads a lot more awkward for me, so I'm a bit hesitant on that part.

**Comment by Gama11** [comment]
Not a native speaker either,  but all the `the` additions seem correct to me. It's not really a matter of preference, without it, it sounds wrong.

File: h2d/Dropdown.hx (hunk 7 above)
Comment (HeapUnderfl0w) [reply h2d/Dropdown.hx:45]: Shouldn't this be `when the user hovers over an item` instead of `when the user overs an item` ?

**Comment by kennygoff** [comment]
Yeah, I'm a native english speaker and the changes are correct with `the`. I'm no linguist but it definitely sounds right and checking some articles seems to confirm that: https://owl.purdue.edu/owl/general_writing/grammar/using_articles.html

**Comment by Yanrishatum** [comment]
As to not spam with Issues when I find questionable places while writing docs, I'm just going to note them here and ping @ncannasse for comments. :)

https://github.com/HeapsIO/heaps/blob/master/h2d/col/IPoint.hx#L32-L38
Is it supposed to return `Point` and not `IPoint` or copy-paste bug?

**Comment by ncannasse** [comment]
Yes it's definitely a copy-paste bug :D

**Comment by Yanrishatum** [comment]
I'll be temporarily hosting more up-to-date API docs on my server in case anyone needs it. https://wb.yanrishatum.ru/heaps/api/

**Comment by ncannasse** [comment]
Here's some comments from a first look:

General

- there's no need IMHO for each class/type to quote it's name at the beginning of the documentation. For instance https://wb.yanrishatum.ru/heaps/api/h2d/Graphics.html , I think it will be better with "~h2d.Graphics provide~ Simple interface to draw arbitrary **2D** geometry." ; Maybe that's not what we used to do, but I think it's better while we're at it to makes things consistent. Also avoid "Is" or "Is used to" or "Provide", or "Implements" as a starting word. It should directly tell what it does, just like we do for methods and variables.

- for variable with default values, the default is already generated in documentation (see https://wb.yanrishatum.ru/heaps/api/h2d/Console.html, both static and shortKeyChar for example). No need then to have the default specified in text documentation. This will also prevent that when changing the default value we break the documentation.

Specific

- I'm not at all a fan of ascii art for expressing complex visual things such as alignments here https://wb.yanrishatum.ru/heaps/api/h2d/Align.html . These should be done either by linking to a manual page that contains images. For instance in that case, the Align class desc should be "Text alignment rules. For more details on how it's applied, see (this manual page)". We could add also for each field "see (this manual page) for more information".

- Console.add is a macro member function. The documentation generator does not mention this and the documentation should be updated.

- filter Ambient apply a color matrix using a mask to tell which parts are to be changed or kept (0...255) - invert tells if the mask should be the parts with ambient applied or the parts that are kept untouched (does 1.0-mask on the mask values) and power allows to control transition speed - pow(mask,power)

Otherwise I don't think I'll be reviewing all of the documentation before merging, as it's often better to have some documentation than none :)

**Comment by Yanrishatum** [comment]
I'm marking this as ready, but I would suggest leaving it open for a bit (a week or so?), in case some crazy enough people could go over it and provide suggestions on grammar and still unclear areas. I won't do as massive doc PRs like that anymore and next API documentation updates are going to be smaller scale and target specific areas.
Also, here's current documentation coverage: https://wb.yanrishatum.ru/heaps/api/coverage.html
Other packages are yikes.

**Comment by Yanrishatum** [comment]
Well, looks like nobody gonna review it, so it's up for merge.

**Comment by ncannasse** [comment]
Sorrrrrry it's conflicting :'( can you fix so I can merge ?

**Comment by Yanrishatum** [comment]
@ncannasse Done. Also, pls fix `h3d.shader.AlphaMap` (post https://github.com/HeapsIO/heaps/commit/865bf5b0b79f6c59c33230b0d70854a4dc93e89c ), it doesn't have `input.uv` declared and it broke documentation generator again ;)

**Comment by ncannasse** [comment]
Done ! Thanks !

**Comment by Yanrishatum** [comment]
Just don't forget to update https://heaps.io/api/ ;)

**Comment by ncannasse** [comment]
Just triggered a rebuild of heaps-doc on Travis, let's see...

**Comment by ncannasse** [comment]
Ok I actually need to make a new heaps release, and update HEAPS_VER
version on heaps.io repo

**Pull request merged**